Add booth limit

nasia-sam opened the pull request:
Add column booth_limit to conference so admin won't be able to accept more booth requests than a conference can support. Method `maximum_accepted_booths?` tests whether the accepted booths have reached the booth limit or not.

nasia-sam commented:
* admin/conference#edit
![20526493_1808079169209844_587037611_n](https://user-images.githubusercontent.com/24506693/28771477-a3b724ae-75eb-11e7-9b59-a31e20e8e210.png)

* admin booths#index
When booth limit hasn't be set (equals to 0) (link redirects to edit_admin_conference_path)

![20526643_1808079195876508_1494108204_n](https://user-images.githubusercontent.com/24506693/28771505-c9b06198-75eb-11e7-87c4-535e04d19d7f.png)

When booth limit is set

![20524324_1808079165876511_1791173859_o](https://user-images.githubusercontent.com/24506693/28771610-2757bd96-75ec-11e7-94f0-fa93aadb7581.png)

When accepted booths are equal to the booth limit and admin tries to exceed it

![20542893_1808079199209841_2117141045_o](https://user-images.githubusercontent.com/24506693/28771665-56664454-75ec-11e7-8d68-a227f7f29963.png)

differentreality commented on app/views/admin/booths/index.html.haml:
```diff
@@ -9,6 +9,22 @@
+        %p
+          Set the
+          = link_to 'Booth limit', edit_admin_conference_path(@conference.short_title)
+          to make sure you are not accepting more booth requests than possible.
```
what about this wording "to make sure you are not accepting more booths than you can accommodate" ?

differentreality commented:
If I cannot accept the booth, why show the option as active in the drop down menu in the first place?

We should also think about the functionality a bit.  

What if I start accepting booths, and suddenly I am no longer able to do so (because of the limit)? 
What am I supposed to do then? 
- If I am supposed to just go and increase the limit, then this basically is not offering us any functionality. 
- If I need to go reject accepted booths, that's not good either.

So now that we have added the existence of booth_limit we need to make good use of it, in a way that it is helpful to the organizer and adds value to our app (If I still need to manually count the booths I want to accept, for instance, then the feature is not adding any extra value).

differentreality commented on app/views/admin/conferences/edit.html.haml:
```diff
@@ -26,4 +26,7 @@
         = f.input :registration_limit, as: :number, in: 0..9999, hint: 'Limit the number of registrations to the conference (0 no limit). Please note that the registration limit doesn\'t apply to speakers of confirmed events (they will still be able to register even if it has been reached). You currently have ' + pluralize(@conference.registrations.count, 'registration')
+      = f.inputs name: 'Booths' do
+        = f.input :booth_limit, as: :number, in: 0..9999,
+        hint: 'Booth limit is the maximum number of booth requets that you can accept for this conference. By setting this number (0 no limit) you can be sure that you are not going to accept more requests than the number of which your conference can support. You currently have ' + pluralize(@conference.booths.where(state: 'accepted').count, 'accepted booth') +'.'
```
* of booths that you can accept..
* ...that you are not going to accept more booths than the conference can accommodate...

differentreality commented on app/views/admin/booths/index.html.haml:
```diff
@@ -9,6 +9,22 @@
+          %b
+            = @conference.booth_limit
+          booth requests. (
+          = pluralize(@conference.booths.where(state: 'accepted').count, 'accepted booth')
```
This could be moved into a scope so that we can simply do @conference.booths.accepted

differentreality commented on app/views/admin/booths/index.html.haml:
```diff
@@ -9,6 +9,22 @@
+          = pluralize(@conference.booths.where(state: 'accepted').count, 'accepted booth')
+          so far)
+      - else
+        %p You have reached the maximum number of accepted booths.
```
What about this also shows the booth_limit number and has a link to the booth_limit config?

differentreality commented on app/controllers/admin/booths_controller.rb:
```diff
@@ -49,7 +51,12 @@ def destroy
 
     def accept
-      update_state(:accept, 'Booth accepted!')
+      if !maximum_accepted_booths?
```
Instead of using negation in your if statement, why not switch the contents of if and else?
So that this reads
```
if maximum_accepted_booths?
  redirect ...
else
  update_state()
end
```

differentreality commented on app/controllers/admin/booths_controller.rb:
```diff
@@ -72,6 +79,15 @@ def cancel
+    # ====Returns
+    # * +True+ -> if accepted booths are equal to the booth limit
+    # * +False+ -> Accepted booths have not reached the booth limit
+    def maximum_accepted_booths?
```
Why isn't this a method in the conference model?

nasia-sam commented:
@differentreality in booths#index I have added a message that informs the organizer about the accepted booths and how many more booths they can accept. However, if you think this is not enough, what if we added confirmation for the accept action where organizer is informed of the remaining booths that are possible to be accepted. I can't think of something else more straightforward. Should I go with that or do you have any other idea ?

differentreality commented:
Please spend some time to think it over, and document your ideas to improve this.

differentreality commented on app/models/conference.rb:
```diff
@@ -739,6 +739,15 @@ def hours_from_start_time(start_hour, end_hour)
+  # * +True+ -> if accepted booths are equal to the booth limit
+  # * +False+ -> Accepted booths have not reached the booth limit
+  def maximum_accepted_booths?
+    booth_limit > 0 && booths.accepted.count + booths.where(state: 'confirmed').count >= booth_limit
```
Why aren't you using the confirmed scope here?!

nasia-sam commented:
* Disabled accept action when the accepted booths equal the booth limit
* Add confirmation for the accept action when booth limit is set 

![20668505_1817813234903104_1279088591_n](https://user-images.githubusercontent.com/24506693/29067861-14698e1a-7c3d-11e7-97e3-da68348e2d2d.png)

* Make messages at booths#index that inform the organizer about the booth limit larger so they attract organizer's attention

differentreality commented:
@nasia-sam please rebase, so that I can review!

differentreality commented on app/views/admin/booths/index.html.haml:
```diff
@@ -9,8 +9,30 @@
             = link_to 'Add Booth', new_admin_conference_booth_path(@conference.short_title), class: 'button btn btn-primary'
       %p.text-muted
         All the booth requests
+
```
One empty line should be enough :)

differentreality commented on app/views/admin/booths/_change_state_dropdown.html.haml:
```diff
@@ -1,7 +1,9 @@
-  %li= link_to 'Accept booth',
-  accept_admin_conference_booth_path(@conference.short_title, booth),
-  method: :patch, id: "accept_booth_#{booth.id}"
+  - unless @conference.maximum_accepted_booths?
```
This should be added in admin_ability.rb, and then here you can simply do `if can? :accept, booth`

differentreality commented on app/views/admin/booths/index.html.haml:
```diff
@@ -9,8 +9,30 @@
+          (
+          = link_to "#{@conference.booth_limit} booths", edit_admin_conference_path(@conference.short_title)
+          )
     .margin-booth-table
```
What's that class doing?

differentreality commented on app/controllers/admin/booths_controller.rb:
```diff
@@ -47,7 +47,9 @@ def update
 
     def accept
-      @booth.accept!
+      if can? :accept, @booth
```
Why do you need to check this?

differentreality commented on app/models/admin_ability.rb:
```diff
@@ -120,6 +125,10 @@ def signed_in_with_organizer_role(user, conf_ids_for_organization_admin = [])
     can :manage, Registration, conference_id: conf_ids
     can :manage, RegistrationPeriod, conference_id: conf_ids
     can :manage, Booth, conference_id: conf_ids
+    cannot :accept, Booth do |booth|
```
Why do we need to repeat this block in all three places?

nasia-sam replied:
All the three of them have the `can :manage, Booth` ability

differentreality replied:
common_abilities_for_roles (https://github.com/nasia-sam/osem/blob/cab940207e346bbf1b957abdfad7217854f25f2c/app/models/admin_ability.rb#L20) doesn't have the can :manage, Booth definition, yet it has the cannot :accept, Booth definition (https://github.com/nasia-sam/osem/blob/cab940207e346bbf1b957abdfad7217854f25f2c/app/models/admin_ability.rb#L78)

More importantly the can :manage, Booth definition depend on other parameters, which are unique to the individual function they appear in. For example, when you are in the organizer role (https://github.com/nasia-sam/osem/blob/cab940207e346bbf1b957abdfad7217854f25f2c/app/models/admin_ability.rb#L109) your definition is:
`can :manage, Booth, conference_id: conf_ids`
https://github.com/nasia-sam/osem/blob/cab940207e346bbf1b957abdfad7217854f25f2c/app/models/admin_ability.rb#L127

But when you are in the cpf role (https://github.com/nasia-sam/osem/blob/cab940207e346bbf1b957abdfad7217854f25f2c/app/models/admin_ability.rb#L178) your definition is:
` can :manage, Booth, conference_id: conf_ids_for_cfp`
https://github.com/nasia-sam/osem/blob/cab940207e346bbf1b957abdfad7217854f25f2c/app/models/admin_ability.rb#L186

So in one case we have the conf_ids_for_cfp and in the other we have conf_ids, each of those two variables hold different values, as such we need to define the can :manage, Booth ability in each function (once for the organizer role, one for the cfp role).

However in all places your definition reads exactly the same, and it does not take into account the role of the user.
```
cannot :accept, Booth do |booth|
      conference = booth.conference
      conference.maximum_accepted_booths?
    end
```

So my question remains:
- Why do we  need to define the exact same block 3 times?
- And based on your previous answer, why is it even defined in common_abilities where there is no can :manage, Booth definition at all?

nasia-sam replied:
I see my mistake. I was testing it with a non-admin user and when the block was written in `signed_in_with_organizer_role`, the   `can :manage, Booth, conference_id: conf_ids_for_cfp` seemed to overwrite it (even if the block I wrote doesn't take into account of the user's role).

I defined the block in the common_abilities section because, although there is no `can :manage, Booth`, there is `can :manage, :all if user.is_admin`. (https://github.com/nasia-sam/osem/blob/cab940207e346bbf1b957abdfad7217854f25f2c/app/models/admin_ability.rb#L54)
Only by putting the block there I could prevent an admin of accepting a booth when `booth_limit ` is reached. 

Taking that into consideration, I will leave the definition of the block in `common_abilities_for_roles` as it seems to be the only section that prevents admins, organizers, _and_ cfps from accepting a booth when they shouldn't.

differentreality replied:
That's right. And moreover it is about the order that things are defined. Even if there was no admin and no admin abilities, common_abilities call the other functions for the abilities for organizer, cfp role etc. If you wanted something (the same thing) to apply for all roles, you could put it in one place, as long as it was **after** their abilities definition, ie at the end of common_abilities.

nasia-sam replied:
I got it now- thanks for the clarifications:)
